hw/misc/aspeed_hace.c

```diff
@@ -104,6 +104,97 @@ static int do_hash_operation(AspeedHACEState *s, int algo)
     return 0;
 }
 
+static int do_hash_sg_operation(AspeedHACEState *s, int algo)
+{
+    hwaddr src, len, dest, buf_len;
+    struct iovec iov[32];
+    unsigned int i = 0;
+    unsigned int is_last = 0;
+    uint8_t *digest_buf = NULL;
+    size_t digest_len = 0;
+    struct aspeed_sg_list *sg_list;
+    int rc;
+    hwaddr check_len = 0;
+
+    src = s->regs[R_HASH_SRC];
+    len = s->regs[R_HASH_SRC_LEN];
+    dest = s->regs[R_HASH_DEST];
+
+    while (!is_last) {
+        buf_len = sizeof(struct aspeed_sg_list);
+        sg_list = (struct aspeed_sg_list *) address_space_map(&s->dram_as,
+                                                      src + (i * buf_len),
+                                                          &buf_len, false,
+                                                   MEMTXATTRS_UNSPECIFIED);
+        if (!sg_list) {
+            qemu_log_mask(LOG_GUEST_ERROR,
+             "%s: failed to map dram for source register (entry %u)\n",
+             __func__, i);
+            return -EACCES;
```

Comment:
these calls are done through memory ops on registers and so returning an errno is useless. It won't be returned to the caller (OS driver)

Reply:
I kept the -EACCESS because Joel's code is doing the same. Even if the guest is completely unaware, I'm guessing Qemu can use that to log failures? Otherwise, please advise how we should return in case of errors.